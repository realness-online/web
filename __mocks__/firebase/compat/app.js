export default {
  initializeApp: vi.fn(),
  apps: []
}
