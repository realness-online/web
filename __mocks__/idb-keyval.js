module.exports.get = jest.fn(() => Promise.resolve(undefined))
module.exports.set = jest.fn(() => Promise.resolve())
module.exports.del = jest.fn(() => Promise.resolve())
module.exports.clear = jest.fn(() => Promise.resolve())
module.exports.keys = jest.fn(() => Promise.resolve([]))
