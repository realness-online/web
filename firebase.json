{
  "storage": {
    "rules": "storage.rules"
  },
  "hosting": {
    "public": "dist",
    "ignore": [
      "firebase.json",
      "**/.*",
      "**/node_modules/**"
    ],
    "rewrites": [
      {
        "source": "**",
        "destination": "/index.html"
      }
    ],
    "headers": [
      {
        "source": "**",
        "headers": [
          {
            "key": "Content-Security-Policy",
            "value": "script-src 'unsafe-eval' 'unsafe-inline' 'self' https://www.google.com https://apis.google.com https://www.gstatic.com https://storage.googleapis.com/"
          }
        ]
      },
      {
        "source": "**",
        "headers": [
          {
            "key": "Cache-Control",
            "value": "max-age=31536000"
          }
        ]
      },
      {
        "source": "/service.worker.js",
        "headers": [
          {
            "key": "Cache-Control",
            "value": "no-cache"
          }
        ]
      }
    ]
  }
}
