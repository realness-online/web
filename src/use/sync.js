import hash from 'object-hash'
import { location, metadata } from '@/use/serverless'
import { get, del, set, keys } from 'idb-keyval'
import {
  as_filename,
  as_author,
  load,
  build_local_directory
} from '@/use/itemid'
import { Offline, Statements, Events, Poster, Me } from '@/persistance/Storage'
import { current_user } from '@/use/serverless'
import { get_my_itemid, use_me } from '@/use/people'
import { use as use_statements } from '@/use/statements'
import get_item from '@/use/item'
import {
  ref,
  onMounted as mounted,
  onUnmounted as dismount,
  nextTick as next_tick,
  getCurrentInstance as current_instance,
  provide,
  watch
} from 'vue'

export const three_minutes = 180000
export const five_minutes = 300000
export const one_hour = 3600000
export const eight_hours = one_hour * 8
export const timeouts = []
export const hash_options = { encoding: 'base64', algorithm: 'md5' }
export const does_not_exist = { updated: null, customMetadata: { md5: null } } // Explicitly setting null to indicate that this file doesn't exist

export const use = () => {
  const { emit } = current_instance()
  const { me, relations } = use_me()
  const { my_statements: statements } = use_statements()
  const events = ref(null)
  const sync_element = ref(null)
  const sync_poster = ref(null)
  provide('sync-poster', sync_poster)
  const play = async () => {
    if (!current_user.value) return // Do nothing until there is a person
    if (document.visibilityState !== 'visible') return
    await visit()
    if (!navigator.onLine || !current_user.value) return
    if (i_am_fresh()) {
      await sync_offline_actions()
    } else {
      emit('active', true)
      // localStorage.sync_time = new Date().toISOString()
      await sync_anonymous_posters()
      await prune()
      await sync_me()
      await sync_statements()
      await sync_events()
      emit('active', false)
    }
  }
  const visit = async () => {
    const visit_digit = new Date(me.value.visited).getTime()
    if (visit_interval() > visit_digit) {
      me.value.visited = new Date().toISOString()
      await next_tick()
      await new Me().save()
    }
  }
  const prune = async () => {
    console.log('prune')
    const everything = await keys()
    everything.forEach(async itemid => {
      if (!as_author(itemid)) return // items have authors
      if (await is_stranger(as_author(itemid), relations.value))
        await del(itemid) // only relations are cached
      if (await itemid.endsWith('/')) await del(itemid)
      else {
        const network = await fresh_metadata(itemid)
        if (!network || !network.customMetadata) return null
        const md5 = await local_md5(itemid)
        if (network.customMetadata.md5 !== md5) await del(itemid)
      }
    })
  }
  const is_stranger = async id => {
    const friends = [...relations.value]
    friends.push({
      id: localStorage.me,
      type: 'person'
    })
    const is_friend = friends.some(relation => {
      if (relation.id === id) return true
      else return false
    })
    return !is_friend
  }
  const sync_offline_actions = async () => {
    if (navigator.onLine) {
      const offline = await get('sync:offline')
      if (!offline) return
      while (offline.length) {
        const item = offline.pop()
        if (item.action === 'save') await new Offline(item.id).save()
        else if (item.action === 'delete') await new Offline(item.id).delete()
        else console.info('weird:unknown-offline-action', item.action, item.id)
      }
      await del('sync:offline')
    }
  }
  const sync_me = async () => {
    console.log('sync_me')
    const id = get_my_itemid()
    const network = (await fresh_metadata(id)).customMetadata
    let my_info = localStorage.getItem(id)
    if (!my_info) my_info = await get(id)
    if (!my_info || !network) return
    const md5 = hash(my_info, hash_options)
    if (md5 !== network.md5) {
      localStorage.removeItem(id)
      del(id)
    }
  }
  const sync_statements = async () => {
    console.log('sync_statements')
    const persistance = new Statements()
    const itemid = get_my_itemid('statements')
    const network = (await fresh_metadata(itemid)).customMetadata
    const elements = sync_element.value.querySelector(`[itemid="${itemid}"]`)
    if (!elements || !elements.outerHTML) return null // nothing local so we'll let it load on request
    const md5 = hash(elements.outerHTML, hash_options)
    if (!network || network.md5 !== md5) {
      statements.value = await persistance.sync()
      if (statements.value.length) {
        await next_tick()
        await persistance.save(elements)
        localStorage.removeItem('/+/statements')
      }
    }
    await persistance.optimize()
  }
  const sync_events = async () => {
    console.log('sync_events')
    const events = new Events()
    const itemid = get_my_itemid('events')
    const network = (await fresh_metadata(itemid)).customMetadata
    const elements = sync_element.value.querySelector(`[itemid="${itemid}"]`)
    if (!elements) return
    const md5 = hash(elements.outerHTML, hash_options)
    if (!network || network.md5 !== md5) {
      events.value = await events.sync()
      if (events.value.length) {
        await next_tick()
        await events.save(elements)
        localStorage.removeItem('/+/events')
      }
    }
  }
  const sync_anonymous_posters = async () => {
    await del('/+/posters/') // TODO:  Maybe overkill
    const offline_posters = await build_local_directory('/+/posters/')
    if (!offline_posters || !offline_posters.items) return
    for (const created_at of offline_posters.items) {
      await save_poster(created_at)
    }
  }
  const save_poster = async created_at => {
    const poster_string = await get(`/+/posters/${created_at}`)
    sync_poster.value = get_item(poster_string)
    sync_poster.value.id = `${localStorage.me}/posters/${created_at}`
    await next_tick()
    await new Poster(sync_poster.value.id).save()
    sync_poster.value = null
    await del(`/+/posters/${created_at}`)
  }
  mounted(async () => {
    document.addEventListener('visibilitychange', play)
    window.addEventListener('online', play)
    relations.value = await load(`${localStorage.me}/relations`)
  })
  dismount(() => {
    window.removeEventListener('online', play)
    document.removeEventListener('visibilitychange', play)
  })
  watch(current_user, async () => {
    if (current_user.value) await play()
  })
  return {
    events,
    sync_element,
    sync_poster
  }
}

export const local_md5 = async itemid => {
  // always checks the network
  const local = await get(itemid)
  if (!local) return null
  return hash(local, hash_options)
}
export const fresh_metadata = async itemid => {
  const index = (await get('sync:index')) || {}
  const path = location(as_filename(itemid))
  let network
  try {
    // console.info('request:metadata', itemid)
    network = await metadata(path)
  } catch (e) {
    if (e.code === 'storage/object-not-found') {
      network = does_not_exist
    } else throw e
  }
  if (!network) throw new Error(`Unable to create metadata for ${itemid}`)
  index[itemid] = network
  await set('sync:index', index)
  return network
}
export function visit_interval() {
  return Date.now() - one_hour
}
export const i_am_fresh = () => {
  let synced
  if (localStorage.sync_time) {
    synced = Date.now() - new Date(localStorage.sync_time).getTime()
  } else {
    localStorage.sync_time = new Date().toISOString()
    synced = eight_hours
  }
  const time_left = eight_hours - synced
  const am_i_fresh = time_left > 0
  console.info('i_am_fresh', am_i_fresh, time_left)
  return am_i_fresh
}
