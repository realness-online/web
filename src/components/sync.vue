<template lang="html">
  <div ref="sync" hidden>
    <as-days v-if="show_statements" v-slot="thoughts"
             itemscope
             :itemid="itemid('statements')"
             :statements="statements">
      <thought-as-article v-for="thought in thoughts"
                          :key="thought[0].id"
                          :statements="thought" />
    </as-days>
    <events-list v-if="events" :events="events" :itemid="itemid('events')" />
    <unsynced-posters v-for="poster in posters"
                      :key="poster.id"
                      :immediate="true"
                      :itemid="poster.id"
                      :poster="poster" />
  </div>
</template>
<script>
  import * as firebase from 'firebase/app'
  import 'firebase/auth'
  import hash from '@/modules/hash'
  import { set, get, del } from 'idb-keyval'
  import { as_type, as_created_at, list } from '@/helpers/itemid'
  import get_item from '@/modules/item'
  import { from_e64 } from '@/helpers/profile'
  import as_days from '@/components/as-days'
  import as_list from '@/components/events/as-list'
  import as_svg from '@/components/posters/as-svg'
  import thought_as_article from '@/components/statements/as-article'
  import {
    Events,
    Statements,
    Poster,
    Offline
  } from '@/persistance/Storage'
  export default {
    components: {
      'as-days': as_days,
      'events-list': as_list,
      'thought-as-article': thought_as_article,
      'unsynced-posters': as_svg
    },
    props: {
      statement: {
        type: Object,
        required: false,
        default: null
      }
    },
    data () {
      return {
        syncer: new Worker('/sync.worker.js'),
        syncing: false,
        person: null,
        posters: null,
        statements: [],
        events: null
      }
    },
    computed: {
      show_statements () {
        if (this.statement) return true
        else if (this.syncing) return true
        return false
      }
    },
    watch: {
      statement () {
        if (this.statement) this.save_statement()
      }
    },
    mounted () {
      firebase.auth().onAuthStateChanged(this.sync_local_storage)
      window.addEventListener('online', this.online)
    },
    beforeDestroy () {
      this.syncer.terminate()
    },
    methods: {
      async save_statement () {
        const itemid = this.itemid('statements')
        this.statements = await list(itemid)
        this.statements.push(this.statement)
        const data = new Statements()
        await this.$nextTick()
        await data.save()
        this.$emit('update:statement', null)
      },
      itemid (type) {
        if (type) return `${localStorage.me}/${type}`
        else return `${localStorage.me}`
      },
      online () {
        this.sync_local_storage(firebase.auth().currentUser)
      },
      async sync_local_storage (current_user) {
        if (navigator.onLine && current_user) {
          console.time('syncing:local-storage')
          this.syncing = true
          localStorage.me = from_e64(current_user.phoneNumber)
          // await Promise.all([
          //   this.sync_statements(),
          //   this.sync_offline(),
          //   this.sync_anonymous_posters(current_user),
          //   this.sync_events()
          // ])
          this.sync_offline()
          this.sync_anonymous_posters(current_user)
          this.sync_events()
          this.sync_statements()
          this.syncing = false
          console.timeEnd('syncing:local-storage')
        }
      },
      async sync_offline () {
        const offline = await get('offline')
        if (!offline) return
        while (offline.length) {
          const item = offline.pop()
          if (item.action === 'save') await new Offline(item.id).save()
          else if (item.action === 'delete') await new Offline(item.id).delete()
          else console.info('unknown offline action', item.action, item.id)
        }
        await del('offline')
      },
      async sync_anonymous_posters (my) {
        const offline_posters = await get('/+/posters/')
        if (!offline_posters || !offline_posters.items) return
        const posters = []
        await Promise.all(offline_posters.items.map(async (created_at) => {
          const poster_as_string = await get(`/+/posters/${created_at}`)
          const poster = get_item(poster_as_string)
          poster.id = `${from_e64(my.phoneNumber)}/posters/${created_at}`
          posters.push(poster)
        }))
        if (posters.length) {
          this.posters = posters
          await this.$nextTick()
          await Promise.all(this.posters.map(async (poster) => {
            const created_at = as_created_at(poster.id)
            const new_poster = new Poster(poster.id)
            await new_poster.save()
            await del(`/+/posters/${created_at}`)
            offline_posters.items = offline_posters.items.filter(when => {
              return parseInt(when) !== created_at
            })
            await set('/+/posters/', offline_posters)
          }))
          this.posters = []
        }
      },
      async sync_events () {
        const itemid = this.itemid('events')
        const events = new Events()
        this.events = await events.sync()
        await this.$nextTick()
        await this.sync_paged(itemid, events)
        this.events = null
      },
      async sync_statements () {
        const itemid = this.itemid('statements')
        const statements = new Statements()
        this.statements = await statements.sync()
        await this.$nextTick()
        await this.sync_paged(itemid, statements)
        this.statements = []
      },
      async sync_paged (itemid, paged) {
        const query = `[itemid="${itemid}"]`
        const elements = this.$el.querySelector(query)
        if (elements) {
          let index = await get('hash-index')
          if (!index) index = {}
          const current_hash = parseInt(index[itemid])
          const new_hash = hash(elements.outerHTML)
          if (current_hash !== new_hash) {
            await paged.save()
            localStorage.removeItem(`/+/${as_type(itemid)}`)
            index[itemid] = new_hash
            set('hash-index', index)
          }
        }
      }
    }
  }
</script>
