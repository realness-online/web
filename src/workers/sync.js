
import * as firebase from 'firebase/app'
import 'firebase/auth'
import 'firebase/storage'
import { is_today } from '@/helpers/date'
import { get, del, keys } from 'idb-keyval'
import { as_type, as_filename } from '@/helpers/itemid'
import { from_e64 } from '@/helpers/profile'
import get_item from '@/modules/item'
import { Offline } from '@/persistance/Storage'

export function message_listener (message) {
  const me = firebase.auth().currentUser
  switch (message.data.action) {
    case 'sync:initialize':
      return initialize(message.data.env)
    case 'sync:people':
      return people(me, message.data.check_everyone)
    case 'sync:offline':
      return offline(me)
    case 'sync:anonymous':
      return anonymous(me)
    default:
      console.warn('Unhandled message from app: ', message.data.action, message)
  }
}
self.addEventListener('message', message_listener) // set up communictaion

export function initialize (credentials) {
  firebase.initializeApp({
    apiKey: credentials.VUE_APP_API_KEY,
    authDomain: credentials.VUE_APP_AUTH_DOMAIN,
    databaseUrl: credentials.VUE_APP_DATABASE_URL,
    projectId: credentials.VUE_APP_PROJECT_ID,
    storageBucket: credentials.VUE_APP_STORAGE_BUCKET,
    messagingSenderId: credentials.VUE_APP_MESSAGING_SENDER_ID
  })
  firebase.auth().onAuthStateChanged(me => {
    if (navigator.onLine && me) {
      people(me, true)
      offline(me)
    }
  })
}
export async function offline (me = firebase.auth().currentUser) {
  if (navigator.onLine && me) {
    const offline = await get('offline')
    while (offline.length) {
      const item = offline.pop()
      if (item.action === 'save') await new Offline(item.id).save()
      else if (item.action === 'delete') await new Offline(item.id).delete()
      else console.info('weird:unknown-offline-action', item.action, item.id)
    }
    await del('offline')
  }
}
export async function anonymous (me = firebase.auth().currentUser) {
  if (navigator.onLine && me) {
    const offline_posters = await get('/+/posters/')
    if (!offline_posters.items) return
    await Promise.all(offline_posters.items.map(async (created_at) => {
      const poster_as_string = await get(`/+/posters/${created_at}`)
      const poster = get_item(poster_as_string)
      poster.id = `${from_e64(me.phoneNumber)}/posters/${created_at}`
      self.postMessage({ action: 'save:poster', poster }, '*')
      await del(`/+/posters/${created_at}`)
    }))
    await del('/+/posters/')
  }
}
export async function people (me = firebase.auth().currentUser, check_everyone = false) {
  if (navigator.onLine && me) {
    console.time('sync:people')
    const people = await list_people()
    await check_people(people, check_everyone)
    console.timeEnd('sync:people')
    setTimeout(async () => { // Call myself in the future
      const is_peering = await get('sync:peer-connected')
      if (!is_peering) await people()
    }, five_seconds)
  }
}

// Local functions
async function list_people () {
  const full_list = await keys()
  return full_list.filter(id => {
    return as_type(id) === 'person'
  })
}
async function check_people (people, check_everyone) {
  const what_I_know = await get('index')
  people.forEach(async (itemid) => {
    const meta = what_I_know[itemid]
    if (check_everyone) prune_person(itemid, what_I_know)
    else if (is_today(meta.updated)) prune_person(itemid, what_I_know)
    // check when each user last updated
  })
}
async function prune_person (itemid, what_I_know) {
  del(`${itemid}/posters/`) // Remove local cache of posters direcrory
  const statements = `${itemid}/statements`
  const events = `${itemid}/events`

  if (await is_outdated(itemid, what_I_know)) {
    del(itemid)
  }
  if (await is_outdated(statements, what_I_know)) {
    del(statements) // Delete statements index
    del(`${statements}/`)
  }
  if (await is_outdated(events, what_I_know)) {
    del(events)
    del(`${events}/`)
  }
}

const five_seconds = 1000 * 5
// const one_minuite = five_seconds * 12
// const five_minutes = five_seconds * one_minuite // 300000
async function is_outdated (itemid, what_I_know) {
  const path = as_filename(itemid)
  const network = await firebase.storage().ref().child(path).getMetadata()
  const local = what_I_know[itemid]
  if (!local) what_I_know[itemid] = network
  const local_time = new Date(what_I_know[itemid].updated).getTime()
  const network_time = new Date(network.updated).getTime()
  if (network_time > local_time) return true
  else return false
}
