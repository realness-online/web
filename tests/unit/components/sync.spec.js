import { shallow } from 'vue-test-utils'
import flushPromises from 'flush-promises'
import { get, set } from 'idb-keyval'
import sync from '@/components/sync'
import * as firebase from 'firebase/app'
import 'firebase/auth'
describe('@/components/sync', () => {
  let wrapper
  const currentUser = {
    phoneNumber: '+16282281824'
  }
  beforeEach(async () => {
    const onAuthStateChanged = jest.fn(state_changed => {
      state_changed(currentUser)
    })
    jest.spyOn(firebase, 'auth').mockImplementation(_ => {
      return { onAuthStateChanged }
    })
    localStorage.me = `/${currentUser}`
    get.mockImplementation(_ => Promise.resolve([]))
    set.mockImplementation(_ => Promise.resolve(null))
    wrapper = shallow(sync)
    await flushPromises()
  })
  it('Renders sync component', async () => {
    expect(wrapper.element).toMatchSnapshot()
    wrapper.destroy()
  })
  it('Terminates worker on destroy', async () => {
    wrapper.destroy()
  })
  describe('Syncronzing localstorage', () => {
    it.todo('Only checks local storage for current user')
    it.todo('Syncronizes and deletes anonymous content once per session')
    describe('Syncronizing models', () => {
      it.todo('Statements')
      it.todo('Profile')
      it.todo('Events')
      it.todo('Posters')
    })
  })
  describe('Sync worker', () => {
    describe('Syncronzing IndexDB:', () => {
      // The application loads the data
      // the syncronizer deletes what's stale
      it.todo('Keeps track of how old items are')
      it.todo('Checks if a resourse has been updated')
      it.todo('Removes any items that are no longer in sync with what\'s current')
    })
    describe('Frequency:', () => {
      it.todo('Checks every 5 minutes for updates from users who have posted recently')
      it.todo('Check once per session for people who haven\'t posted in a while')
      it.todo('will rely on peer to peer for updates sooner than five minutes')
    })
    describe('Peer syncing:', () => {
      it.todo('contacts a signaling server to determine available peers')
      it.todo('Connects to available peer')
      it.todo('share indexes with peer')
      it.todo('shares updates with peers')
    })
  })
})
