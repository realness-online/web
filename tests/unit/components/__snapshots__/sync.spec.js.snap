// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`@/components/sync Render Renders sync component 1`] = `
<div
  hidden="hidden"
>
  <!---->
   
  <as-days-stub
    avatars=""
    events=""
    itemid="/+16282281824/statements"
    itemscope=""
    posters=""
    statements=""
  />
   
  <!---->
   
  <!---->
</div>
`;
